### ISSUE TITLE
from_iter_instead_of_collect suggests invalid replacement

### ISSUE THREAD

**Issue by markbt** [rust-lang/rust-clippy#6648]
Lint name: **from_iter_instead_of_collect**

The suggestion provided by this lint is incorrect if the parameter is a range.  It suggests something like `0..3.collect()`, however the `.collect()` binds to the end of the range.  Parentheses must be used to get the right syntax.  It is also missing the type hint that must be necessarily added (either via the turbofish or by specifying the type of the target variable if this is a variable assignment).

Minimal example:

```rust
use std::iter::FromIterator;

fn main() {
    let a = vec![0, 1, 2];
    assert_eq!(a, Vec::from_iter(0..3));

    // Clippy suggests this, which does not compile:
    // assert_eq!(a, 0..3.collect());

    // This would be the correct suggestion:
    assert_eq!(a, (0..3).collect::<Vec<_>>());
}
```

The lint output looks like this:

```
warning: usage of `FromIterator::from_iter`
 --> src/main.rs:5:19
  |
5 |     assert_eq!(a, Vec::from_iter(0..3));
  |                   ^^^^^^^^^^^^^^^^^^^^ help: use `.collect()` instead of `::from_iter()`: `0..3.collect()`
  |
```

### Meta

- `cargo clippy -V`: clippy 0.0.212 (e1884a8e 2020-12-29)
- `rustc -Vv`:
```
rustc 1.49.0 (e1884a8e3 2020-12-29)
binary: rustc
commit-hash: e1884a8e3c3e813aada8254edfa120e85bf5ffca
commit-date: 2020-12-29
host: x86_64-apple-darwin
release: 1.49.0
```
